### PR TITLE
Add non-dispatching listenable action future

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently the only implementation of `ListenableActionFuture` requires
dispatching listener execution to a thread pool. This commit renames
that variant to `DispatchingListenableActionFuture` and converts
`AbstractListenableActionFuture` to be a variant that does not require
dispatching. That class is now named `PlainListenableActionFuture`.
Both variants are created from static methods methods on 
`PlainListenableActionFuture`.

### CODE REVIEW

**Comment by tbrooks8** [comment]
I'm not sure with this change if we want to maintain the `ListenableActionFuture` interface? That interface essentially allows a method to expose a `ListenableActionFuture` that can be read or have a listener added to, but does not exposed the "set" methods.

But there is currently only one implementation of that interface `PlainListenableActionFuture` (the dispatching version extends `PlainListenableActionFuture`).

**Comment by javanna** [comment]
hey @tbrooks8 which "set" method to do you mean is missing? 

Looking at where the `addListener` method is used, only in tests and can be easily replaced by using the execute variant that accepts a listener, I think that it's there only for Java api users. But I am not sure whether people use it in practice and what for. We have an execute method that accepts a listener, and one that returns a future, not sure why we need to have listenable futures, I think that it makes the api confusing (see #9201). Not sure about the interface either, we expose it only in `ActionRequestBuilder#execute` as a return type (again java API), and not when using one of the `Client#execute` methods. This seems inconsistent, I would clean it up. I opened #24415 which should make it possible to remove the interface as a followup or do whatever we want to do with it as it becomes internal.

As a side note: in the high level REST client we went for one method that accepts a listener, and one that returns the response straight-away. No futures nor listenable futures.

Out of curiosity, what do we need `DispatchingListenableActionFuture` for?

**Comment by tbrooks8** [comment]
@javanna 

So essentially what I mean is: the interface `ListenableActionFuture` extends `ActionFuture`. So right now, an a method can return this interface. This interface defines the ability to "block waiting for the result" and "add listener to be executed on completion.

I was saying, as the only implementation of this interface after my PR is `PlainListenableActionFuture`, maybe we do not need the interface. But the downside of that is `PlainListenableActionFuture` extends `AdapterActionFuture` which exposes the `onResponse` and `onFailure` methods which are used to complete the future. So you could no longer have a method that returned a `ListenableActionFuture` without exposing a way to complete the future.

> Out of curiosity, what do we need DispatchingListenableActionFuture for?

I'm not sure. After your PR, there may not be a need. The only other place I see it used it could be replaced by a `PlainActionFuture`. I introduced this PR because I need a non-dispatching listenable future for transport stuff. But after these PRs land there may not be a need for the dispatching version anymore?

File: core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 1]
```diff
@@ -19,17 +19,110 @@
...
+    volatile Object listeners;
+    boolean executedListeners = false;
+
+    private PlainListenableActionFuture() {}
+
+    public static <T> PlainListenableActionFuture<T> newListenableFuture() {
+        return new PlainListenableActionFuture<>();
+    }
+
+    public static <T> PlainListenableActionFuture<T> newDispatchingListenableFuture(ThreadPool threadPool) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java]: can you add some javadocs to these public statics?

File: core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 2]
```diff
@@ -19,17 +19,110 @@
...
+        try {
+            // we use a timeout of 0 to by pass assertion forbidding to call actionGet() (blocking) on a network thread.
+            // here we know we will never block
+            listener.onResponse(actionGet(0));
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
     }
 
+    private static class DispatchingListenableActionFuture<T> extends PlainListenableActionFuture<T> {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java]: make the class final?

File: core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 3]
```diff
@@ -19,17 +19,110 @@
...
+        private DispatchingListenableActionFuture(ThreadPool threadPool) {
+            this.threadPool = threadPool;
+        }
+
+        @Override
+        public void addListener(final ActionListener<T> listener) {
+            super.addListener(new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.LISTENER, listener, false));
+        }
+
+        public ThreadPool threadPool() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java]: the clas is private so we can make this method private too?

**Review by s1monw — APPROVED** [review]
left some minor nit picks

File: core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java (hunk 3 above)
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java]: I actually do not think that is necessary anymore, so I will get rid of it.

**Review by s1monw — APPROVED** [review]
LGTM

**Comment by javanna** [comment]
heya @tbrooks8 with my recent changes, I see usages of `ListenableActionFuture` and `PlainListenableActionFuture` only in their own tests. We could potentially remove them from our codebase as far as I can see but given that you will need something similar in the future, up to you what you want to do with those depending on your plans.

File: core/src/test/java/org/elasticsearch/client/transport/TransportClientRetryIT.java [hunk 4]
```diff
@@ -69,7 +69,7 @@ public void testRetry() throws IOException, ExecutionException, InterruptedExcep
                 if (randomBoolean()) {
                     clusterState = client.admin().cluster().state(clusterStateRequest).get().getState();
                 } else {
-                    PlainListenableActionFuture<ClusterStateResponse> future = new PlainListenableActionFuture<>(client.threadPool());
+                    PlainActionFuture<ClusterStateResponse> future = PlainActionFuture.newFuture();
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/client/transport/TransportClientRetryIT.java]: oops, I didn't realize I stole these changes from you. I happened to make exactly the same changes to these two transport tests. Sorry!

**Pull request merged**